# xsnp
Cross-Sample SNP Analysis Tools

The input refers to the pileup file generated by the [mapping.rules](https://github.com/czbiohub/sunbeam_gt-pro/blob/master/rules/mapping/mapping.rules), by mapping shotgun metagenomic reads to a large number of samples and species.   

The output is a compact representation of the input that can be loaded into R to enable downstream analyses with PCA, tSNE, clusterig, tree building, etc.

See [Data Schema And Computation](https://github.com/czbiohub/xsnp/wiki/Data-Schema-And-Computation) and [Roadmap](https://github.com/czbiohub/xsnp/wiki/Roadmap)

## Usage

First pass:

```
xsnp/scripts/go.py list_of_pileup_files
```

Second pass:
```
cd banded

for sample in `cat ../hmp_all.txt`; do for x in {0..47}; do echo ${sample%.pileup}.sites.gcb10.dp2.band${x}.tsv >> band${x}.txt; done; done

for i in {0..47}; do echo band${i}.txt; xsnp/scripts/go2.py "band${i}".txt; done
```
